web-components/chart/chart.tsx

```diff
@@ -0,0 +1,306 @@
+import { Component, Prop, Listen, Element } from '@stencil/core';
+import * as Vega from 'vega';
+import * as VegaLite from 'vega-lite';
+import vegaTooltip from 'vega-tooltip';
+import { csv as d3csv } from 'd3-fetch';
+
+@Component({
+  tag: 'cob-chart',
+  styleUrl: 'chart.css',
+})
+export class CobChart {
+  @Element() el;
+
+  @Prop({ mutable: true })
+  config: any;
+  chartID: string = '';
+  minWidth: number = 0;
+  view: any;
+  compiledSpec: any;
+  dataset: any;
+  selectField: string = '';
+  selectOptions: any;
+  vegaLite: boolean = true;
+
+  // We listen for a window resize and adjust the width of the chart if it happens
+  @Listen('window:resize')
+  setChartWidth() {
+    // We use the chart div to set the width of the svg chart element and
+    // subtract 50px from it to account for padding.
+    const newWidth = this.el.getBoundingClientRect().width - 50;
+
+    // We don't want our chart's width to be less than the minWidth,
+    // so if the new wrapper div width is less than that, we set it
+    // manually.
+    this.view.width(newWidth >= this.minWidth ? newWidth : this.minWidth).run();
+
+    // If we're using vega and we're specifically building a pie chart then
+    // we want to make sure the width of the chart doesn't exceed the height
+    // to prevent the pie from getting cut off.
+    if (
+      this.vegaLite == false &&
+      this.config.marks.filter(item => item.type == 'arc').length >= 1
+    ) {
+      // In this case we have a max width which is the given height of the chart
+      const maxWidth = this.config.height;
+      this.view.width(newWidth >= maxWidth ? maxWidth : newWidth).run();
+    }
+
+    // We can't use "autosize" on facet charts in Vega, so if we're using columns
+    // we update the "child_width" and "child_height" signals to update
+    // the size of the chart.
+    if (
+      this.vegaLite == true &&
+      typeof this.config.encoding.column !== 'undefined'
+    ) {
+      /* The "width" in facet/grouped charts refers to the width of 
+        each column in the chart, so we:
+        1. calculate the number of columns in the chart
+        2. get the width of the y-axis
+        3. calculate the appropriate width for each column
+      */
+
+      // We get number of columns in the chart by calucating the unique
+      // values in the field being used to group the data.
+      const field = this.config.encoding.column.field;
+      const numColumns = [...new Set(this.dataset.map(item => item[field]))]
```

Comment:
`this.dataset.length` gives me a count of all the records in the dataset, but I want to know the number of unique values in a particular field so I know how many columns the chart has. Set seems useful since it does the de-duplicating for me. I cleaned up the line a little bit and expanded the comment, but lmk if there is a better way to do this.